File: tts/translator.go
package tts

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

func NeedTranslate(effects string) (string, bool) {
	effSlice := strings.Split(effects, "-")
	for i, eff := range effSlice {
		if strings.ToLower(eff) == "tr" {
			effSlice[i] = effSlice[len(effSlice)-1]
			effSlice = effSlice[:len(effSlice)-1]
			return strings.Join(effSlice, "-"), true
		}
	}
	return effects, false
}

func getTranslateReq(lang, text string) string {
	v := url.Values{}
	v.Add("client", "gtx")
	v.Add("dt", "t")
	v.Add("sl", "auto")
	v.Add("tl", lang)
	v.Add("q", text)
	return "https://translate.googleapis.com/translate_a/single?" + v.Encode()
}

func Translate(lang, text string) (string, error) {
	reqUrl := getTranslateReq(lang, text)

	resp, err := httpClient.Get(reqUrl)
	if err != nil {
		return "", err
	}

	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("status not ok: %d", resp.StatusCode)
	}
	var result []any

	limitData := io.LimitReader(resp.Body, 8<<10)

	data, err := io.ReadAll(limitData)
	if err != nil {
		return "", fmt.Errorf("can't read data: %v", err)
	}

	err = json.Unmarshal(data, &result)
	if err != nil {
		return "", fmt.Errorf("can't decode data: %v", err)
	}
	if len(result) == 0 {
		return "", errors.New("empty translate response")
	}
	var finalResult string

	firstLevel, ok := result[0].([]any)
	if !ok {
		return "", errors.New("bad json structure")
	}
	for i := range firstLevel {
		secondLevel, ok := firstLevel[i].([]any)
		if !ok {
			continue
		}
		result, ok := secondLevel[0].(string)
		if !ok {
			continue
		}
		finalResult += result
	}

	return finalResult, nil
}

File: tts/Translator.cs
using System.Net;
using System.Text.Json;
using System.Web;

namespace TwitchIrcMinimal;

public static class Translator
{
    private static readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(5) };

    public static (string effects, bool translated) NeedTranslate(string? effects)
    {
        if (string.IsNullOrEmpty(effects)) return (effects ?? "", false);
        var parts = effects.Split('-', StringSplitOptions.RemoveEmptyEntries);
        for (int i = 0; i < parts.Length; i++)
        {
            if (parts[i].Equals("tr", StringComparison.OrdinalIgnoreCase))
            {
                var result = parts.Where((_, idx) => idx != i).ToArray();
                return (string.Join("-", result), true);
            }
        }
        return (effects, false);
    }

    public static async Task<string?> TranslateAsync(string lang, string text)
    {
        var query = HttpUtility.ParseQueryString("");
        query["client"] = "gtx";
        query["dt"] = "t";
        query["sl"] = "auto";
        query["tl"] = lang;
        query["q"] = text;
        var url = $"https://translate.googleapis.com/translate_a/single?{query}";

        try
        {
            var resp = await _http.GetAsync(url);
            if (resp.StatusCode != HttpStatusCode.OK) return null;

            var json = await resp.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.GetArrayLength() == 0) return null;

            var firstLevel = root[0];
            var result = "";
            for (int i = 0; i < firstLevel.GetArrayLength(); i++)
            {
                var segment = firstLevel[i];
                if (segment.GetArrayLength() > 0)
                    result += segment[0].GetString() ?? "";
            }
            return result;
        }
        catch { return null; }
    }
}
